connect/runtime/src/test/java/org/apache/kafka/connect/integration/ExactlyOnceSourceIntegrationTest.java

```diff
@@ -0,0 +1,997 @@
+                            offsetsTopic
+                    ).iterator().next();
+            long seqno = parseAndAssertOffsetForSingleTask(offsetRecord);
+            assertEquals("Offset commits should occur on connector-defined poll boundaries, which happen every " + recordsProduced + " records",
+                    0, seqno % recordsProduced);
+
+            // also consume from the cluster's global offsets topic; again, just need to read one offset record
+            offsetRecord = connect.kafka()
+                    .consume(
+                            1,
+                            TimeUnit.MINUTES.toMillis(1),
+                            globalOffsetsTopic
+                    ).iterator().next();
+            seqno = parseAndAssertOffsetForSingleTask(offsetRecord);
+            assertEquals("Offset commits should occur on connector-defined poll boundaries, which happen every " + recordsProduced + " records",
+                    0, seqno % recordsProduced);
+
+            // Shut down the whole cluster
+            connect.workers().forEach(connect::removeWorker);
+            // Reconfigure the cluster with exactly-once support disabled
+            workerProps.put(EXACTLY_ONCE_SOURCE_SUPPORT_CONFIG, "disabled");
+
+            // Establish new expectations for records+offsets
+            connectorHandle.expectedRecords(recordsProduced);
+            connectorHandle.expectedCommits(recordsProduced);
+
+            // Restart the whole cluster
+            for (int i = 0; i < DEFAULT_NUM_WORKERS; i++) {
+                connect.addWorker();
+            }
+
+            // And perform a basic sanity check that the cluster is able to come back up, our connector and its task are able to resume running,
+            // and the task is still able to produce source records and commit offsets
+            connect.assertions().assertAtLeastNumWorkersAreUp(DEFAULT_NUM_WORKERS, "cluster did not restart in time");
+            connect.assertions().assertConnectorAndExactlyNumTasksAreRunning(
+                    CONNECTOR_NAME,
+                    1,
+                    "connector and tasks did not resume running after cluster restart in time"
+            );
+
+            log.info("Waiting for records to be provided to worker by task");
+            // wait for the connector tasks to produce enough records
+            connectorHandle.awaitRecords(SOURCE_TASK_PRODUCE_TIMEOUT_MS);
+
+            log.info("Waiting for records to be committed to Kafka by worker");
+            // wait for the connector tasks to commit enough records
+            connectorHandle.awaitCommits(TimeUnit.MINUTES.toMillis(1));
+
+            // Delete the connector before shutting down the Kafka cluster it's targeting
+            connect.deleteConnector(CONNECTOR_NAME);
+        }
+    }
+
+    @Test
+    public void testPotentialDeadlockWhenProducingToOffsetsTopic() throws Exception {
```

Comment:
Could we have a comment explaining the potential deadlock?

Reply:
Ack, done